File: simulator/xbee.py
#!/usr/bin/python3
import serial
import queue
import time
import binascii

# class for managing XBee API mode 
class XBee():
    # Bytes which need to be escaped in frames
    SPECIAL_BYTES = {
        'FRAME_DELIM':  0x7E,
        'ESCAPE':       0x7D,
        'XON':          0x11,
        'XOFF':         0x13,
    }

    # Digimesh broadcast addr
    BROADCAST = 0x000000000000FFFF

    # Digimesh frame types
    FRAME_TYPES = {
        'AT':           0x08,
        'AT_QPV':       0x09,
        'TX':           0x10,
        'EXPLICIT_TX':  0x11,
        'REMOTE':       0x17,
        'AT_RESP':      0x88,
        'MODEM_STATUS': 0x8A,
        'TX_STATUS':    0x8B,
        'ROUTE_INFO':   0x8D,
        'RX':           0x90,
        'EXPLICIT_RX':  0x91,
        'NODE_ID':      0x95,
        'REMOTE_RESP':  0x97,
    }

    BUF_FULL_TIMEOUT = 5

    # queue to hold validated frames
    rx_queue = queue.Queue()
    # buffer to hold partial frames
    rx_buf = bytearray()

    def __init__(self, devfile, baud):
        self.serial = serial.Serial(port=devfile, 
                                    baudrate=baud,
                                    parity=serial.PARITY_NONE,
                                    stopbits=serial.STOPBITS_ONE,
                                    bytesize=serial.EIGHTBITS
                                    )
        self.serial.flush()
        self.max_frame = 255

    def tx(self, data, dest=0x000000000000FFFF, opts=0x00):
        '''
        data is an unescaped string.
        dest is the 64-bit digimesh address to tx to.
        opts are the frame options.
        '''
        frame_size = len(data) + 14     # tx api frame has 14 bytes overhead
        if (frame_size > self.max_frame):
            print("Frame too long, not sending.")
            return False

        frame = bytearray(((frame_size >> 8) & 0x0FF, 
                           (frame_size & 0x0FF),
                           self.FRAME_TYPES['TX'],
                           0x00,
                           # destination
                           (dest >> 56) & 0x0FF,
                           (dest >> 48) & 0x0FF,
                           (dest >> 40) & 0x0FF,
                           (dest >> 32) & 0x0FF,
                           (dest >> 24) & 0x0FF,
                           (dest >> 16) & 0x0FF,
                           (dest >> 8) & 0x0FF,
                           (dest & 0x0FF),
                           0xFF, # reserved
                           0xFE, # reserved
                           0x00, # broadcast radius (default 0x00 for radius=max hops)
                           0x00, # tx options (default, use whatever is already set)
                           ))

        # append the data
        frame += bytearray(data.encode())

        # append checksum
        frame.append(0xFF - (sum(frame[2:]) & 0x0FF))
        
        # escape the frame
        frame = self.escape(frame)

        # prepend the unescaped delimiter
        frame = bytearray(b'\x7E') + frame

        return self.serial.write(frame)

    def rx(self):
        '''
        Read bytes from serial if any are waiting.
        Then validate frames and add to rx_queue.
        '''
        received = False
        num_bytes = self.serial.in_waiting
        self.rx_buf.extend(self.serial.read(num_bytes))
        # each sequence is a frame. split at the delimiter.
        sequence = self.rx_buf.split(bytes(b'\x7E'))        
        for s in sequence:
            frame = self.validate_frame(s)
            if (frame is not None):
                self.rx_queue.put(frame, self.BUF_FULL_TIMEOUT)
                received = True

        if (received == True):
            self.rx_buf = bytearray()
            return self.rx_queue.get()

        else:
            return None

    def parse_frame(self, frame):
        '''
        Parse a frame. Figure out what type it is, and return the relevant data.
        '''
        for frametype, value in self.FRAME_TYPES.items():
            if frame[2] == value:
                break;

        if (value == self.FRAME_TYPES['AT_RESP']):
            pass
        elif (value == self.FRAME_TYPES['MODEM_STATUS']):
            pass
        elif (value == self.FRAME_TYPES['TX_STATUS']):
            pass
        elif (value == self.FRAME_TYPES['ROUTE_INFO']):
            pass
        elif (value == self.FRAME_TYPES['RX']):
            source = frame[3:10]
            opts = frame[13]
            data = frame[14:-1]
            return data
        elif (value == self.FRAME_TYPES['EXPLICIT_RX']):
            pass
        elif (value == self.FRAME_TYPES['TX']):
            # just for debugging. the only time this should ever show up is
            # if you have a serial cable hooked up to the TX pin of the xbee.
            pass
        elif (value == self.FRAME_TYPES['NODE_ID']):
            pass
        elif (value == self.FRAME_TYPES['REMOTE_RESP']):
            pass
        else:
            raise Exception("Frame has invalid frame type.")
        return None

    def escape(self, data):
        ''' Put an escape character before a special character, and escape it. '''
        escaped = bytearray()
        for byte in data:
            if byte in self.SPECIAL_BYTES.values():
                escaped.append(self.SPECIAL_BYTES['ESCAPE'])
                escaped.append(byte ^ 0x20)
            else:
                escaped.append(byte)
        return escaped

    def unescape(self, data):
        ''' Think "escape", but the opposite of that. '''
        unescaped = bytearray()
        last_escaped = False
        # create an iterator from range() so we can use next() on it
        q = iter(range(len(data)))
        for i in q:
            if (last_escaped):
                unescaped.append(data[i] ^ 0x20)
                last_escaped = False
            elif data[i] == self.SPECIAL_BYTES['ESCAPE']:
                last_escaped = True
            else:
                unescaped.append(data[i])
        return unescaped

    def validate_frame(self, frame):
        ''' Confirm that a frame is valid. Return None otherwise. '''
        frame = self.unescape(frame)
        if (frame is not None):
            if (len(frame) < 5):
                # already know this isn't a valid frame: not enough overhead
                frame = None
                return None
            else:
                frame_len = (frame[0] << 8) | frame[1]
            # make sure frame length is accurate
            if (len(frame[2:-1]) != frame_len):
                #print('invalid len: ' + str(len(frame[2:-1])) + ' act: ' + str(frame_len))
                frame = None
            # check checksum
            elif ((sum(frame[2:]) & 0x0FF) != 0xFF):
                #print('invalid checksum: ' + '{:02X}'.format(sum(frame[2:]) & 0x0FF))
                #print(frame)
                frame = None
        return frame
                
    def get_frame(self):
        ''' Returns the oldest frame in the queue, or None. '''
        if (self.rx_queue.empty() == False):
            return self.rx_queue.get(block=False)
        else:
            return None
        

if __name__ == '__main__':
    xb = XBee('/dev/xbee', 1200)
    while(True):
        r = xb.rx()
        while (r == None):
            r = xb.rx()
        h = binascii.hexlify(r[16:-1])
        q = [h[i:i+2] for i in range(0, len(h), 2)]
        c = ""
        for d in q:
            c += " " + d.decode()
        print(c)
        print("---")

File: xbee.c
#include "xbee.h"
#include "rbuf.h"
#include "status.h"
#include "serial.h"
#include <avr/io.h>
#include <util/delay.h>

volatile rbuf_t rbuf;
uint8_t func_code = 0x00;
uint8_t err_code = 0x00;

const static struct frame_types_t FRAME_TYPES =
{
    .AT             = (uint8_t)0x08,
    .AT_QPV         = (uint8_t)0x09,
    .TX             = (uint8_t)0x10,
    .EXPLICIT_TX    = (uint8_t)0x11,
    .REMOTE         = (uint8_t)0x17,
    .AT_RESP        = (uint8_t)0x88,
    .MODEM_STATUS   = (uint8_t)0x8A,
    .TX_STATUS      = (uint8_t)0x8B,
    .ROUTE_INFO     = (uint8_t)0x8D,
    .RX             = (uint8_t)0x90,
    .EXPLICIT_RX    = (uint8_t)0x91,
    .NODE_ID        = (uint8_t)0x95,
    .REMOTE_RESP    = (uint8_t)0x97
};

const static struct special_bytes_t SPECIAL_BYTES =
{
    .FRAME_DELIM = (uint8_t)0x7E,
    .ESCAPE      = (uint8_t)0x7D,
    .XON         = (uint8_t)0x11,
    .XOFF        = (uint8_t)0x13
};

void xbee_init()
{
    rbuf.start = 0;
    rbuf.end = 0;
    //TX_INT_ENABLE();
    RX_INT_ENABLE();
}

uint8_t tx(uint8_t *data, uint8_t dsize, uint64_t dest, uint8_t opts)
{
    uint8_t frame[MAX_BUF_SIZE];

    // TX frame has 14 bytes overhead
    // does not include delimiter or length
    uint8_t fsize = dsize + 14;
    uint8_t sum = 0;

    // delim + len_high + len_low + fsize + checksum
    // must fit in the buffer.
    if ((fsize + 5) > MAX_BUF_SIZE)
        return FRAME_SIZE_ERR;

    frame[0] = 0x7E;
    frame[1] = (uint8_t)(fsize >> 8);
    frame[2] = (uint8_t)(fsize);
    frame[3] = FRAME_TYPES.TX;
    frame[4] = 0x01;
    frame[5] = (uint8_t)(dest >> 56);
    frame[6] = (uint8_t)(dest >> 48);
    frame[7] = (uint8_t)(dest >> 40);
    frame[8] = (uint8_t)(dest >> 32);
    frame[9] = (uint8_t)(dest >> 24);
    frame[10] = (uint8_t)(dest >> 16);
    frame[11] = (uint8_t)(dest >> 8);
    frame[12] = (uint8_t)(dest);
    frame[13] = 0xFF;   // reserved
    frame[14] = 0xFE;   // reserved
    frame[15] = 0x00;   // broadcast radius (default 0x00 for radius=max hops)
    frame[16] = opts;   // tx options (probably should just be 0x00)

    // compute first part of checksum.
    for (int i=3; i<17; i++)
    {
        sum += frame[i];
    }

    // append data and sum it
    for (int i=0; i<dsize; i++)
    {
        frame[17 + i] = data[i];
        sum += data[i];
    }
    // put checksum at the end
    frame[fsize + 3] = 0xFF - sum;

    // send it
    cli();
    for (int i=0; i < fsize + 5; i++)
        put_byte(frame[i]);
    sei();
    return 0;
}

//! rx(frame) assumes frame has MAX_BUF_SIZE bytes allocated already.
//! DO NOT use this if frame is unallocated.
uint8_t rx(uint8_t *frame)
{
    uint8_t ret;
    // Add timeout here.
    do
    {
        ret = find_frame(&rbuf, frame);
    }
    while (ret != 0);
    return ret;
}

// This will shift up to the frame delimiter, or shift
// everything out if one was not found.
void shift_to_delim(volatile rbuf_t *r)
{
    if (rbuf_read(r, 0) != SPECIAL_BYTES.FRAME_DELIM)
    {
        // Find the first frame delimiter.
        for (uint16_t i=1; i<MAX_BUF_SIZE; i++)
        {
            if (rbuf_read(r, i) == SPECIAL_BYTES.FRAME_DELIM)
            {
                rbuf_shift(r, i);
                break;
            }
        }
    }
}

//! Checks the receive buffer for any potential frames. 
//! Try validation, and then shift out of the buffer if validated.
//! It is important that no interrupts call rbuf_shift()
//! while this function is executing.
uint8_t find_frame(volatile rbuf_t *r, uint8_t *frame)
{
    uint16_t buf_len;
    uint8_t ret;
    // Check that the first byte is a frame delimiter.
    // If not, shift out bytes until we hit one.
    shift_to_delim(r);

    if (rbuf_read(r, 0) == SPECIAL_BYTES.FRAME_DELIM)
    {
        buf_len = rbuf_len(r);

        for (int i=0; i < buf_len; i++)
        {
            frame[i] = rbuf_read(r, i);
        }

        unescape(frame, MAX_BUF_SIZE);
        ret =  validate_frame(frame, MAX_BUF_SIZE);
    }
    else
    {
        // could not find frame delimiter.
        ret = FRAME_DELIM_ERR;
    }
    return ret;
}

//! Loops through the frame, unescaping any escaped bytes.
//! Could be done in find_frame and save a loop, but let's see if
//! that's necessary before premature optimization...
void unescape(uint8_t *frame, uint16_t size)
{
    uint16_t i = 1;
    uint16_t j;
    // stop if we reach the end of the array. 
    while (i < size)
    {
        // Check that we reached an escape byte.
        if (frame[i] == SPECIAL_BYTES.ESCAPE)
        {
            j = i;
            // shift the right side of byte array to the left.
            while (j+1 < size)
            {
                frame[j] = frame[j+1];
                j++;
            }

            // Unescape the character.
            frame[i] = frame[i] ^ 0x20;
        }
        i++;
    }
}

uint8_t escape(uint8_t *frame, uint16_t size)
{
    uint16_t i = 1;
    uint16_t j;

    while (i < size)
    {
        if (frame[i] == SPECIAL_BYTES.FRAME_DELIM ||
            frame[i] == SPECIAL_BYTES.ESCAPE      ||
            frame[i] == SPECIAL_BYTES.XON         ||
            frame[i] == SPECIAL_BYTES.XOFF)
        {
            if (i + 1 == size) // can't add an escape char, buffer full
                return FRAME_SIZE_ERR;
            j = size - 1;
            // shift right side of array to the right.
            while (j > i)
            {
                frame[j] = frame[j - 1];
                j--;
            }
            frame[i] = SPECIAL_BYTES.ESCAPE;
            frame[i+1] = frame[i+1] ^ 0x20;
            i++; // make sure we increment i twice in this case.
        }
        i++;
    }
    return 0;
}

//! Check the checksum. 
uint8_t validate_frame(uint8_t *frame, uint16_t size)
{
    uint8_t ret = 0;
    uint8_t sum = 0;
    uint16_t data_len, frame_len, buf_len;

    // check that we have at least frame_len # of bytes in the buffer.
    data_len = ((uint16_t)frame[1] << 8) | (uint16_t)frame[2];
    frame_len = data_len + 4;
    buf_len = rbuf_len(&rbuf);
    if (frame_len > buf_len)
    {
        if (frame_len > MAX_BUF_SIZE)
        {
            // Frame too large for the buffer.
            ret = FRAME_SIZE_ERR;
            shift_to_delim(&rbuf);
        }
        else
        {
            ret = FRAME_RX_INCOMPLETE;
        }
    }
    else
    {
        // Sum from byte 3 to the end of the frame, including checksum.
        for (int i=3; i < (frame_len); i++)
        {
            sum += frame[i];
        }
        // Make sure they add to 0xFF.
        if ((uint8_t)(sum & 0xFF) != (uint8_t)0xFF)
        {
            ret = FRAME_SUM_ERR;
        }
        // Shift it out of the buffer, whether it's good or not.
        shift_to_delim(&rbuf);
    }
    // possibly make this nicer at some point...
    return ret;
}

ISR(USART_RX_vect)
{
    uint8_t tmp = UDR0;
    rbuf_append(&rbuf, tmp);
}

